hwilib/devices/ledger.py

```diff
@@ -117,209 +147,272 @@ def func(*args: Any, **kwargs: Any) -> Any:
-                if is_p2wsh(scriptcode):
-                    if len(psbt_in.witness_script) == 0:
-                        continue
-                    scriptcode = psbt_in.witness_script
-                elif is_p2wpkh(scriptcode):
-                    _, _, wit_prog = is_witness(scriptcode)
-                    scriptcode = b"\x76\xa9\x14" + wit_prog + b"\x88\xac"
+                if p2sh:
+                    if wit_ver == 0:
+                        script_addrtype = AddressType.SH_WIT
+                    else:
+                        raise BadArgumentError("Cannot have witness v1+ in p2sh")
                 else:
+                    if wit_ver == 0:
+                        script_addrtype = AddressType.WIT
+                    elif wit_ver == 1:
+                        script_addrtype = AddressType.TAP
+                    else:
+                        continue
+
+            # Check if P2WSH
+            if is_p2wsh(scriptcode):
+                if len(psbt_in.witness_script) == 0:
                     continue
-                has_segwit = True
+                scriptcode = psbt_in.witness_script
+
+            multisig = parse_multisig(scriptcode)
+            if multisig is not None:
+                k, ms_pubkeys = multisig
+
+                # Figure out the parent xpubs
+                key_exprs: List[str] = []
+                ok = True
+                our_keys = 0
+                for pub in ms_pubkeys:
+                    if pub in psbt_in.hd_keypaths:
+                        pk_origin = psbt_in.hd_keypaths[pub]
+                        if pk_origin.fingerprint == master_fp:
+                            our_keys += 1
+                            pubkeys[input_num] = pub
+                        for xpub_bytes, xpub_origin in psbt2.xpub.items():
+                            xpub = ExtendedKey.from_bytes(xpub_bytes)
+                            if (xpub_origin.fingerprint == pk_origin.fingerprint) and (xpub_origin.path == pk_origin.path[:len(xpub_origin.path)]):
+                                key_exprs.append(PubkeyProvider(xpub_origin, xpub.to_string(), "/**").to_string())
+                                break
+                        else:
+                            # No xpub, Ledger will not accept this multisig
+                            ok = False
+                    else:
+                        # No origin info, Ledger will not accept this multisig
+                        ok = False
+
+                if not ok:
+                    retry_legacy = False
+                    continue
+
+                if our_keys != 1:
+                    # Ledger does not allow having more than one key per multisig
+                    retry_legacy = False
+                    continue
+
+                # Make and register the MultisigWallet
+                msw = MultisigWallet(f"{k} of {len(key_exprs)} Multisig", script_addrtype, k, key_exprs)
+                _, registered_hmac = self.client.register_wallet(msw)
```

Comment:
While the `wallets` variable is a set and therefore each added entry is different even if there are multiple inputs for the same wallet, the `register_wallet` call requires user's interaction. It would be nice to not repeat it for the same wallet, by checking if there is an entry in `wallets` with the same policy map as `msw` before registering.

Reply:
Done.